File: app/src/main/java/com/boardgamegeek/ui/model/PlayStats.java
package com.boardgamegeek.ui.model;

import java.util.ArrayList;
import java.util.List;

public class PlayStats {
	public static final int INVALID_FRIENDLESS = Integer.MIN_VALUE;
	public static final double INVALID_UTILIZATION = -1.0;
	public static final double INVALID_CFM = -1.0;

	private int numberOfPlays = 0;
	private int numberOfPlayedGames = 0;
	private int numberOfQuarters = 0;
	private int numberOfDimes = 0;
	private int numberOfNickels = 0;
	private int gameHIndex = 0;
	private List<HIndexEntry> hIndexGames = new ArrayList<>();
	private int friendless = INVALID_FRIENDLESS;
	private double utilization = INVALID_UTILIZATION;
	private double cfm = 0.0;
	private int top100count = 0;

	private PlayStats(Builder builder) {
		numberOfPlays = builder.numberOfPlays;
		numberOfPlayedGames = builder.numberOfPlayedGames;
		numberOfQuarters = builder.numberOfQuarters;
		numberOfDimes = builder.numberOfDimes;
		numberOfNickels = builder.numberOfNickels;
		gameHIndex = builder.gameHIndex;
		hIndexGames = builder.hIndexGames;
		friendless = builder.friendless;
		utilization = builder.utilization;
		cfm = builder.cfm;
		top100count = builder.top100count;
	}

	public int getNumberOfPlays() {
		return numberOfPlays;
	}

	public int getNumberOfGames() {
		return numberOfPlayedGames;
	}

	public int getNumberOfQuarters() {
		return numberOfQuarters;
	}

	public int getNumberOfDimes() {
		return numberOfDimes;
	}

	public int getNumberOfNickels() {
		return numberOfNickels;
	}

	public int getGameHIndex() {
		return gameHIndex;
	}

	public int getFriendless() {
		return friendless;
	}

	public double getUtilization() {
		return utilization;
	}

	public double getCfm() {
		return cfm;
	}

	public int getTop100Count() {
		return top100count;
	}

	public List<HIndexEntry> getHIndexGames() {
		return hIndexGames;
	}

	public static final class Builder {
		private int numberOfPlays;
		private int numberOfPlayedGames;
		private int numberOfQuarters;
		private int numberOfDimes;
		private int numberOfNickels;
		private int gameHIndex;
		private List<HIndexEntry> hIndexGames = new ArrayList<>();
		private int friendless;
		private double utilization;
		private double cfm;
		private int top100count;

		public Builder() {
		}

		public Builder numberOfPlays(int val) {
			numberOfPlays = val;
			return this;
		}

		public Builder numberOfPlayedGames(int val) {
			numberOfPlayedGames = val;
			return this;
		}

		public Builder numberOfQuarters(int val) {
			numberOfQuarters = val;
			return this;
		}

		public Builder numberOfDimes(int val) {
			numberOfDimes = val;
			return this;
		}

		public Builder numberOfNickels(int val) {
			numberOfNickels = val;
			return this;
		}

		public Builder gameHIndex(int val) {
			gameHIndex = val;
			return this;
		}

		public Builder hIndexGames(List<HIndexEntry> val) {
			hIndexGames = val;
			return this;
		}

		public Builder friendless(int val) {
			friendless = val;
			return this;
		}

		public Builder utilization(double val) {
			utilization = val;
			return this;
		}

		public Builder cfm(double val) {
			cfm = val;
			return this;
		}

		public Builder top100count(int val) {
			top100count = val;
			return this;
		}

		public PlayStats build() {
			return new PlayStats(this);
		}
	}
}

File: app/src/main/java/com/boardgamegeek/model/PlayStats.kt
package com.boardgamegeek.model

import com.boardgamegeek.extensions.cdf
import com.boardgamegeek.extensions.inverseCdf
import kotlin.math.ln

class PlayStats private constructor(private val games: List<GameForPlayStats>, private val ownedQuantity: Int = OWNED_QUANTITY_UNKNOWN) {
    private var _hIndex: HIndex = HIndex.invalid()
    val hIndex: HIndex
        get() = _hIndex

    private var _gIndex: GIndex = GIndex.invalid()
    val gIndex: GIndex
        get() = _gIndex

    val numberOfPlays: Int by lazy {
        games.sumOf { it.playCount }
    }

    val numberOfPlayedGames: Int by lazy {
        games.filter { it.playCount > 0 }.size
    }

    val numberOfNickels: Int by lazy {
        games.filter { it.playCount in 5..9 }.size
    }

    val numberOfDimes: Int by lazy {
        games.filter { it.playCount in 10..24 }.size
    }

    val numberOfQuarters: Int by lazy {
        games.filter { it.playCount in 25..49 }.size
    }

    val numberOfHalfDollars: Int by lazy {
        games.filter { it.playCount in 50..99 }.size
    }

    val numberOfDollars: Int by lazy {
        games.filter { it.playCount >= 100 }.size
    }

    val top100Count: Int by lazy {
        games.filter { it.playCount > 0 }.filter { it.bggRank in 1..100 }.size
    }

    fun getHIndexGames(): List<Pair<String, Int>> {
        if (hIndex.h == 0) return emptyList()
        return games.map { it.name to it.playCount }.sortedByDescending { it.second }
    }


    val friendless: Int by lazy {
        val numberOfOwnedGamesThatHaveEarnedTheirKeep = ownedAndPlayedGames.filter { it.playCount >= PLAY_COUNT_TO_EARN_KEEP }.size
        val numberOfOwnedUnplayedGames = ownedQuantity - numberOfOwnedAndPlayedGames
        when {
            ownedQuantity == OWNED_QUANTITY_UNKNOWN -> INVALID_FRIENDLESS
            numberOfOwnedAndPlayedGames == 0 -> 0
            numberOfOwnedGamesThatHaveEarnedTheirKeep >= ownedAndPlayedGames.size -> ownedAndPlayedGames.minOfOrNull { it.playCount } ?: INVALID_FRIENDLESS
            else -> {
                val index = ownedAndPlayedGames.lastIndex - numberOfOwnedGamesThatHaveEarnedTheirKeep
                val friendless = ownedAndPlayedGames.sortedByDescending { it.playCount }.getOrNull(index)?.playCount ?: INVALID_FRIENDLESS
                if (friendless == 0) numberOfOwnedGamesThatHaveEarnedTheirKeep - numberOfOwnedUnplayedGames else friendless
            }
        }
    }

    val cfm: Double by lazy {
        // continuous Friendless metric
        when {
            ownedQuantity == OWNED_QUANTITY_UNKNOWN -> INVALID_CFM
            numberOfOwnedAndPlayedGames == 0 -> 0.0
            else -> (totalCdf / numberOfOwnedAndPlayedGames).inverseCdf(lambda)
        }
    }

    val utilization: Double by lazy {
        // “the amount of novelty you've gained from that game compared to all the novelty that can ever be had”
        // every 10 plays utilizes another 90%
        when {
            ownedQuantity == OWNED_QUANTITY_UNKNOWN -> INVALID_UTILIZATION
            numberOfOwnedAndPlayedGames == 0 -> 0.0
            else -> totalCdf / ownedQuantity
        }
    }

    private val ownedAndPlayedGames: List<GameForPlayStats> by lazy {
        games.filter { it.isOwned }
    }

    private val numberOfOwnedAndPlayedGames: Int by lazy {
        ownedAndPlayedGames.size
    }

    private val totalCdf: Double by lazy {
        ownedAndPlayedGames.sumOf { it.playCount.toDouble().cdf(lambda) }
    }

    companion object {
        const val OWNED_QUANTITY_UNKNOWN = -1
        const val INVALID_FRIENDLESS = Integer.MIN_VALUE
        const val INVALID_UTILIZATION = -1.0
        const val INVALID_CFM = -1.0
        private const val PLAY_COUNT_TO_EARN_KEEP = 10
        private val lambda = ln(0.1) / -10

        suspend fun fromList(games: List<GameForPlayStats>, ownedQuantity: Int): PlayStats {
            return PlayStats(games, ownedQuantity).also {
                val playCounts = games.map { game -> game.playCount }
                it._hIndex = HIndex.fromList(playCounts)
                it._gIndex = GIndex.fromList(playCounts)
            }
        }
    }
}
